File: app/src/main/java/com/ch/yoon/kakao/pay/imagesearch/ui/imagesearch/ImageSearchActivity.java
package com.ch.yoon.kakao.pay.imagesearch.ui.imagesearch;

import android.content.Intent;
import android.os.Bundle;
import android.view.KeyEvent;

import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.ch.yoon.kakao.pay.imagesearch.R;
import com.ch.yoon.kakao.pay.imagesearch.databinding.ActivityImageSearchBinding;
import com.ch.yoon.kakao.pay.imagesearch.repository.ImageRepository;
import com.ch.yoon.kakao.pay.imagesearch.repository.ImageRepositoryImpl;
import com.ch.yoon.kakao.pay.imagesearch.repository.local.room.ImageDatabase;
import com.ch.yoon.kakao.pay.imagesearch.repository.local.room.ImageLocalDataSource;
import com.ch.yoon.kakao.pay.imagesearch.repository.local.room.dao.ImageSearchDao;
import com.ch.yoon.kakao.pay.imagesearch.repository.remote.kakao.ImageRemoteDataSource;
import com.ch.yoon.kakao.pay.imagesearch.ui.base.BaseActivity;
import com.ch.yoon.kakao.pay.imagesearch.ui.imagedetail.ImageDetailActivity;
import com.ch.yoon.kakao.pay.imagesearch.ui.imagesearch.imagelist.ImageListViewModel;
import com.ch.yoon.kakao.pay.imagesearch.ui.imagesearch.imagelist.ImageListViewModelFactory;
import com.ch.yoon.kakao.pay.imagesearch.ui.imagesearch.imagelist.adapter.ImageListAdapter;
import com.ch.yoon.kakao.pay.imagesearch.ui.imagesearch.searchbox.adapter.SearchLogAdapter;
import com.ch.yoon.kakao.pay.imagesearch.ui.imagesearch.imagelist.helper.ImageSearchInspector;
import com.ch.yoon.kakao.pay.imagesearch.ui.imagesearch.searchbox.SearchBoxViewModel;
import com.ch.yoon.kakao.pay.imagesearch.ui.imagesearch.searchbox.SearchBoxViewModelFactory;

public class ImageSearchActivity extends BaseActivity<ActivityImageSearchBinding> {

    private ActivityImageSearchBinding binding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = binding(R.layout.activity_image_search);

        initSearchBoxViewModel();
        initSearchKeywordEditText();
        observeSearchBoxViewModel();

        initImageListViewModel();
        observeImageListViewModel();

        initSearchLogRecyclerView();
        initImageRecyclerView();
    }

    private void initSearchBoxViewModel() {
        final ImageSearchDao imageSearchDao = ImageDatabase.getInstance(getApplicationContext()).imageDocumentDao();
        final ImageLocalDataSource localDataSource = ImageLocalDataSource.getInstance(imageSearchDao);
        final ImageRemoteDataSource remoteDataSource = ImageRemoteDataSource.getInstance();
        final ImageRepository repository = ImageRepositoryImpl.getInstance(localDataSource, remoteDataSource);

        final SearchBoxViewModel viewModel = ViewModelProviders.of(this,
            new SearchBoxViewModelFactory(getApplication(), repository)
        ).get(SearchBoxViewModel.class);

        binding.setSearchBoxViewModel(viewModel);
    }

    private void initSearchKeywordEditText() {
        binding.keywordEditText.setOnEditorActionListener((v, actionId, event) -> {
            if(actionId == KeyEvent.KEYCODE_ENDCALL) {
                String text = v.getText().toString();
                binding.getSearchBoxViewModel().clickSearchButton(text);
                return true;
            }

            return false;
        });
    }

    private void observeSearchBoxViewModel() {
        binding.getSearchBoxViewModel()
            .observeSearchKeyword()
            .observe(this, keyword -> {
                binding.getImageListViewModel().loadImageList(keyword);
                binding.keywordEditText.setText(keyword);
            });

        binding.getSearchBoxViewModel()
            .observeSearchBoxFinish()
            .observe(this, voidEvent ->
                finish()
            );

        binding.getSearchBoxViewModel()
            .observeShowMessage()
            .observe(this, this::showToast);
    }

    private void initSearchLogRecyclerView() {
        final LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this, RecyclerView.VERTICAL, false);
        binding.searchLogRecyclerView.setLayoutManager(linearLayoutManager);

        final SearchLogAdapter adapter = new SearchLogAdapter();
        adapter.setOnSearchLogClickListener((searchLog, position) ->
            binding.getSearchBoxViewModel().clickSearchButton(searchLog.getKeyword())
        );

        adapter.setOnLogDeleteClickListener((searchLog, position) ->
            binding.getSearchBoxViewModel().clickKeywordDeleteButton(searchLog.getKeyword())
        );

        binding.searchLogRecyclerView.setAdapter(adapter);
    }

    private void initImageListViewModel() {
        final ImageSearchDao imageSearchDao = ImageDatabase.getInstance(getApplicationContext()).imageDocumentDao();
        final ImageLocalDataSource localDataSource = ImageLocalDataSource.getInstance(imageSearchDao);
        final ImageRemoteDataSource remoteDataSource = ImageRemoteDataSource.getInstance();
        final ImageRepository repository = ImageRepositoryImpl.getInstance(localDataSource, remoteDataSource);

        final ImageSearchInspector imageSearchInspector = new ImageSearchInspector(1, 50, 80, 20);

        final ImageListViewModel viewModel = ViewModelProviders.of(this, new ImageListViewModelFactory(
            getApplication(), repository, imageSearchInspector
        )).get(ImageListViewModel.class);

        binding.setImageListViewModel(viewModel);
    }

    private void observeImageListViewModel() {
        binding.getImageListViewModel()
            .observeShowMessage()
            .observe(this, this::showToast);
    }

    private void initImageRecyclerView() {
        final ImageListAdapter imageListAdapter = new ImageListAdapter();

        imageListAdapter.setOnBindPositionListener(position ->
            binding.getImageListViewModel().loadMoreImageListIfPossible(position)
        );

        imageListAdapter.setOnListItemClickListener((imageDocument, position) -> {
            Intent imageDetailIntent = ImageDetailActivity.getImageDetailActivityIntent(this, imageDocument);
            startActivity(imageDetailIntent);
        });

        imageListAdapter.setOnFooterItemClickListener(() ->
            binding.getImageListViewModel().retryLoadMoreImageList()
        );

        GridLayoutManager gridLayoutManager = (GridLayoutManager)binding.imageRecyclerView.getLayoutManager();
        if(gridLayoutManager != null) {
            gridLayoutManager.setSpanSizeLookup(new GridLayoutManager.SpanSizeLookup() {
                @Override
                public int getSpanSize(int position) {
                    if(imageListAdapter.isFooterViewPosition(position)) {
                        return gridLayoutManager.getSpanCount();
                    } else {
                        return 1;
                    }
                }
            });
        }

        binding.imageRecyclerView.setLayoutManager(gridLayoutManager);
        binding.imageRecyclerView.setAdapter(imageListAdapter);
    }

    @Override
    public void onBackPressed() {
        binding.getSearchBoxViewModel().clickBackPress();
    }

}

File: app/src/main/java/com/ch/yoon/imagesearch/presentation/search/ImageSearchActivity.kt
package com.ch.yoon.imagesearch.presentation.search

import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.widget.AbsListView
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.ch.yoon.imagesearch.R
import com.ch.yoon.imagesearch.databinding.ActivityImageSearchBinding
import com.ch.yoon.imagesearch.extension.throttleFirstWithOneSecond
import com.ch.yoon.imagesearch.presentation.base.BaseActivity
import com.ch.yoon.imagesearch.presentation.favorite.FavoriteImagesActivity
import com.ch.yoon.imagesearch.presentation.detail.ImageDetailActivity
import com.ch.yoon.imagesearch.presentation.search.backpress.BackPressViewModel
import com.ch.yoon.imagesearch.presentation.search.imagesearch.ImageSearchViewModel
import com.ch.yoon.imagesearch.presentation.search.imagesearch.ImageSearchResultsAdapter
import com.ch.yoon.imagesearch.presentation.search.searchbox.SearchBoxViewModel
import com.jakewharton.rxbinding2.view.clicks
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.subjects.PublishSubject
import org.koin.androidx.viewmodel.ext.android.viewModel
import java.util.concurrent.TimeUnit

/**
 * Creator : ch-yoon
 * Date : 2019-10-27.
 */
class ImageSearchActivity : BaseActivity<ActivityImageSearchBinding>() {

    private val searchBoxViewModel: SearchBoxViewModel by viewModel()
    private val imageSearchViewModel: ImageSearchViewModel by viewModel()
    private val backPressViewModel: BackPressViewModel by viewModel()

    override fun getLayoutId(): Int {
        return R.layout.activity_image_search
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        initActionBar()

        observeBackPressViewModel()
        initSearchBoxViewModel()
        observeSearchBoxViewModel()

        initImageListViewModel()
        observeImageListViewModel()
        initImageRecyclerView()
    }

    private fun initActionBar() {
        setSupportActionBar(binding.mainToolbar)
        supportActionBar?.setDisplayShowTitleEnabled(false)
    }

    private fun observeBackPressViewModel() {
        val owner = this
        with(backPressViewModel) {
            showMessageEvent.observe(owner, Observer { message ->
                showToast(message)
            })

            finishEvent.observe(owner, Observer {
                finish()
            })
        }
    }

    private fun initSearchBoxViewModel() {
        binding.searchBoxViewModel = searchBoxViewModel

        if(isActivityFirstCreate) {
            searchBoxViewModel.loadSearchLogList()
        }
    }

    private fun observeSearchBoxViewModel() {
        val owner = this
        with(searchBoxViewModel) {
            searchEvent.observe(owner, Observer { keyword ->
                imageSearchViewModel.loadImageList(keyword)
            })

            showMessageEvent.observe(owner, Observer { message ->
                showToast(message)
            })
        }
    }

    private fun initImageListViewModel() {
        binding.searchListViewModel = imageSearchViewModel
    }

    private fun observeImageListViewModel() {
        val owner = this
        with(imageSearchViewModel) {
            showMessageEvent.observe(owner, Observer { message ->
                showToast(message)
            })

            moveToDetailScreenEvent.observe(owner, Observer { imageDocument ->
                val imageDetailIntent = ImageDetailActivity.getImageDetailActivityIntent(owner, imageDocument)
                startActivity(imageDetailIntent)
            })
        }
    }

    private fun initImageRecyclerView() {
        binding.imageRecyclerView.apply {

            adapter = ImageSearchResultsAdapter().apply {
                onBindPosition = { position ->
                    imageSearchViewModel.loadMoreImageListIfPossible(position)
                }

                itemClickSubject.throttleFirstWithOneSecond()
                    .subscribe { imageSearchViewModel.onClickImage(it) }
                    .disposeByOnDestroy()

                footerClickSubject.throttleFirstWithOneSecond()
                    .subscribe { imageSearchViewModel.retryLoadMoreImageList() }
                    .disposeByOnDestroy()
            }

            layoutManager = (layoutManager as GridLayoutManager).apply {
                spanSizeLookup = object : GridLayoutManager.SpanSizeLookup() {
                    override fun getSpanSize(position: Int): Int {
                        return if ((adapter as ImageSearchResultsAdapter).isFooterViewPosition(position)) {
                            spanCount
                        } else {
                            1
                        }
                    }
                }
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.image_search_menu, menu)
        menu?.let {
            it.findItem(R.id.action_search).clicks()
                .throttleFirstWithOneSecond()
                .subscribe { searchBoxViewModel.onClickShowButton() }
                .disposeByOnDestroy()

            it.findItem(R.id.action_favorite).clicks()
                .throttleFirstWithOneSecond()
                .subscribe { startActivity(Intent(this, FavoriteImagesActivity::class.java)) }
                .disposeByOnDestroy()
        }
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        return item?.let { menuItem ->
            when(menuItem.itemId) {
                R.id.action_search -> { true }
                R.id.action_favorite -> { true }
                else -> { false }
            }
        } ?: super.onOptionsItemSelected(item)
    }

    override fun onBackPressed() {
        if(searchBoxViewModel.isOpen) {
            searchBoxViewModel.onClickHideButton()
        } else {
            backPressViewModel.onBackPress()
        }
    }
}